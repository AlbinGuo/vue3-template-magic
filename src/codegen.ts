import { NodeTypes } from "./ast";

interface Context {
  code: string;
  push(code: string): void;
}

// TODO ast type from vue3
export function generate(ast: any) {
  const context = createCodegenContext();
  const node = ast.children[0];

  genNode(node, context);

  return {
    code: context.code,
  };
}

function createCodegenContext(): Context {
  const context = {
    code: "",
    push(source: string) {
      context.code += source;
    },
  };

  return context;
}

function genNode(node: any, context: Context) {
  switch (node.type) {
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    default:
      break;
  }
}

function genText(node: any, context: Context) {
  const { push } = context;
  push(`${node.content}`);
}

function genElement(node: any, context: Context) {
  const { push } = context;
  const tagName = node.tag;
  const isSelfClosing = node.isSelfClosing;

  if (isSelfClosing) {
    push(`<${tagName}/>`);
  } else {
    push(`<${tagName}`);
    genElementAttributes(node, context);
    push(`>`);

    if (node.children?.length > 0) {
      genNodeList(node, context);
    }
    push(`</${tagName}>`);
  }
}

function genElementAttributes(node: any, context: Context) {
  if (!node.props || node.props.length === 0) return;

  node.props.forEach((propsNode: any) =>
    genElementAttribute(propsNode, context)
  );
}

function genElementAttribute(node: any, context: Context) {
  const { push } = context;
  let key = node.name;
  let value = "";
  if (node.value) {
    value = `="${node.value.content}"`;
  }
  push(` ${key}${value}`);
}

function genNodeList(node: any, context: Context) {
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    genNode(children[i], context);
  }
}
