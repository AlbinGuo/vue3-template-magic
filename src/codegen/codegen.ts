// TODO 这里的 ast 类型应该是从 vue3 中导出
export function generate(ast: any) {
  const node = ast.children[0];

  const tagName = node.tag;
  const isSelfClosing = node.isSelfClosing;

  let code = "";

  if (isSelfClosing) {
    code += `<${tagName}/>`;
  } else {
    code += `<${tagName}></${tagName}>`;
  }

  return {
    code,
  };
}
