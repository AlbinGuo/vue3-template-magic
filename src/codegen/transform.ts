export function transform(root: any, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
}

function createTransformContext(root: any, options: any): any {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };

  return context;
}

function traverseNode(node: any, context: any) {
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) exitFns.push(onExit);
  }

  switch (node.type) {
    case "root":
    case "element":
      traverseChildren(node, context);
      break;

    default:
      break;
  }

  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}
function traverseChildren(node: any, context: any) {
  const children = node.children;

  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}
