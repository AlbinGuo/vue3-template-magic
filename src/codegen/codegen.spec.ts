import { describe, it, expect } from "vitest";
import { generate } from "./codegen";
import { baseParse } from "@vue/compiler-core";

describe("codegen", () => {
  it("simple element", () => {
    const root = {
      type: "root",
      children: [
        {
          type: "element",
          tag: "div",
        },
      ],
    };

    const { code } = generate(root);
    expect(code).toMatchSnapshot();
  });

  it(" self closing element  ", () => {
    const root = {
      type: "root",
      children: [
        {
          type: "element",
          tag: "span",
          isSelfClosing: true,
        },
      ],
    };

    const { code } = generate(root);
    expect(code).toMatchSnapshot();
  });
});
