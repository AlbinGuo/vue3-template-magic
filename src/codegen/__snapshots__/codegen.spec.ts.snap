// Vitest Snapshot v1

exports[`codegen >  self closing element   1`] = `"<span/>"`;

exports[`codegen > simple element 1`] = `"<div></div>"`;
